contracts/libraries/Assessment/AssessmentVoteLib.sol

```diff
@@ -0,0 +1,176 @@
+    IAssessment.Incident[] storage incidents
+  ) internal view returns (uint) {
+    if (eventType == IAssessment.EventType.CLAIM) {
+      IAssessment.ClaimDetails memory claimDetails = claims[id].details;
+      return claimDetails.amount * CONFIG.REWARD_PERC * claimDetails.coverPeriod / 365 / PERC_BASIS_POINTS;
+    }
+    IAssessment.IncidentDetails memory incidentDetails = incidents[id].details;
+    uint payoutImpact = AssessmentUtilsLib._getPayoutImpactOfIncident(incidentDetails);
+    return payoutImpact * CONFIG.REWARD_PERC / PERC_BASIS_POINTS;
+  }
+
+  // [todo] Expose a view to find out the last index until withdrawals can be made and also
+  //  views for total rewards and withdrawable rewards
+  function withdrawReward (
+    IAssessment.Configuration calldata CONFIG,
+    INXMToken nxm,
+    address user,
+    uint104 untilIndex,
+    mapping(address => IAssessment.Stake) storage stakeOf,
+    mapping(address => IAssessment.Vote[]) storage votesOf,
+    IAssessment.Claim[] storage claims,
+    IAssessment.Incident[] storage incidents
+  ) external returns (uint rewardToWithdraw, uint104 withdrawUntilIndex) {
+    IAssessment.Stake memory stake = stakeOf[user];
+    {
+      uint voteCount = votesOf[user].length;
+      withdrawUntilIndex = untilIndex > 0 ? untilIndex : uint104(voteCount);
+      require(
+        untilIndex <= voteCount,
+        "Vote count is smaller that the provided untilIndex"
+      );
+      require(stake.rewardsWithdrawnUntilIndex < voteCount, "No withdrawable rewards");
+    }
+
+    uint totalReward;
+    for (uint i = stake.rewardsWithdrawnUntilIndex; i < withdrawUntilIndex; i++) {
+      IAssessment.Vote memory vote = votesOf[user][i];
+      require(
+        block.timestamp > AssessmentUtilsLib._getVoteLockupEndDate(CONFIG, vote),
+        "Cannot withdraw rewards from votes which are in lockup period"
+      );
+      IAssessment.Poll memory poll =
+        IAssessment.EventType(vote.eventType) == IAssessment.EventType.CLAIM
+        ? claims[vote.eventId].poll
+        : incidents[vote.eventId].poll;
+
+      totalReward = _getTotalRewardForEvent(
+        CONFIG,
+        IAssessment.EventType(vote.eventType),
+        vote.eventId,
+        claims,
+        incidents
+      );
+      rewardToWithdraw += totalReward * vote.tokenWeight / (poll.accepted + poll.denied);
+    }
+
+    stakeOf[user].rewardsWithdrawnUntilIndex = withdrawUntilIndex;
+    nxm.mint(user, rewardToWithdraw);
+  }
+
+  function castVote (
+    IAssessment.Configuration calldata CONFIG,
+    uint8 eventType,
+    uint104 id,
+    bool accepted,
+    mapping(address => IAssessment.Stake) storage stakeOf,
+    mapping(address => IAssessment.Vote[]) storage votesOf,
+    mapping(bytes32 => bool) storage hasAlreadyVotedOn,
+    IAssessment.Claim[] storage claims,
+    IAssessment.Incident[] storage incidents
+  ) external {
+
+    {
+      bytes32 voteHash = keccak256(abi.encodePacked(id, msg.sender, eventType));
```

Comment:
check if mapping(member => mapping(id => boolOrEventTypeToBoolMapping)) uses 2 SLOADS or just one

Reply:
It doesn't, but the bytecode increases significantly and for some reason it also uses more gas (probably because the hash is made on 4 nonpacked keys?).